Pass sanitizer flags to swift driver

etcwilde opened the pull request:
This patch fixes the part of the build system where we aren't passing
the sanitizer flags to swift, so the linker isn't linking against the
sanitizer libraries.

This should work on Windows. I'm frustrated enough with Windows that I'm just going to call it a night. compnerd, if you see something that is out of place, please let me know.

rdar://79313476

etcwilde committed:
Pass sanitizer flags to swift on Unix-y systems

This patch fixes the part of the build system where we aren't passing
the sanitizer flags to swift, so the linker isn't linking against the
sanitizer libraries.

etcwilde commented:
@swift-ci please ASAN test macOS Platform

varungandhi-apple commented on cmake/modules/AddSwift.cmake:
```diff
@@ -75,6 +75,41 @@ function(_set_target_prefix_and_suffix target kind sdk)
+        if(LLVM_USE_SANITIZER STREQUAL "MemoryWithOrigins")
+          # Not supported
+        endif()
+      elseif (LLVM_USE_SANITIZER STREQUAL "Undefined")
+        set(_Swift_SANITIZER_FLAGS "-sanitize=undefined")
+      elseif (LLVM_USE_SANITIZER STREQUAL "Thread")
+        set(_Swift_SANITIZER_FLAGS "-sanitize=thread")
+      elseif (LLVM_USE_SANITIZER STREQUAL "DataFlow")
+        # Not supported
+      elseif (LLVM_USE_SANITIZER STREQUAL "Address;Undefined" OR
+              LLVM_USE_SANITIZER STREQUAL "Undefined;Address")
+        set(_Swift_SANITIZER_FLAGS "-sanitize=address -sanitize=undefined")
```
Using [`FIND`](https://cmake.org/cmake/help/latest/command/string.html#find) might be simpler compared to trying to match combinations using `STREQUAL`.

etcwilde replied:
Maybe. I just pulled the block from LLVM so that the logic matches.
Eventually, we should upstream it so the LLVM is assigning the flags appropriately, but this should get us through until that happens. I'd like to avoid deviating from what they're doing for now if at all possible.

varungandhi-apple replied:
Didn't realize the context, sounds reasonable.

etcwilde committed:
This should work on Windows

After talking with compnerd, because we're passing this directly to the
swift driver, we shouldn't have to do anything weird with the flags and
it should "just work". I'm frustrated enough with Windows that I'm just
going to go with that and go to bed. If I broke something, I'm sorry. I
tried as far as my sanity would allow.

etcwilde commented:
@swift-ci please test

etcwilde commented:
@swift-ci ASAN test

etcwilde commented:
@swift-ci please test macOS platform

etcwilde commented:
@swift-ci ASAN test

compnerd commented on cmake/modules/AddSwift.cmake:
```diff
@@ -75,6 +75,36 @@ function(_set_target_prefix_and_suffix target kind sdk)
 endfunction()
 
+function(_add_host_variant_swift_sanitizer_flags target)
+  if (LLVM_USE_SANITIZER)
```
I really want to say, please use LLVM style ...

```cmake
  if(NOT LLVM_USE_SANITIZER)
    return()
  endif()
```

That way we can unindent the rest of the implementation.  I'll let you decide on that.

compnerd commented on cmake/modules/AddSwift.cmake:
```diff
@@ -75,6 +75,36 @@ function(_set_target_prefix_and_suffix target kind sdk)
+    elseif (LLVM_USE_SANITIZER STREQUAL "Leaks")
+      # Not supported
+    else()
+      message(FATAL_ERROR "Unsupported value of LLVM_USE_SANITIZER: ${LLVM_USE_SANITIZER}")
```
```suggestion
      message(SEND_ERROR "unsupported value for LLVM_USE_SANITIZER: ${LLVM_USE_SANITIZER}")
```

That way we should be able to finish the configure phase and identify any other issues but prevent further progress.

etcwilde committed:
Update cmake/modules/AddSwift.cmake

Use `SEND_ERROR` instead of `FATAL_ERROR`.

Co-authored-by: Saleem Abdulrasool <compnerd@compnerd.org>

etcwilde committed:
Fix spacing in if's

Little fix for the indentation of the if's.

etcwilde commented:
@swift-ci please test

DougGregor commented:
@swift-ci please test macOS

etcwilde commented:
still timing out. :(

etcwilde commented:
@swift-ci please test macOS

etcwilde commented:
@swift-ci please test macOS

gottesmm commented:
@swift-ci test macOS platform

etcwilde commented:
I'm trying to see if the failure reproduces locally. It doesn't seem like it should be caused by this PR, but just want to verify.

gottesmm commented:
@swift-ci test macOS platform

etcwilde commented:
I'm not getting these errors when testing locally under the same build-it invocation.

etcwilde commented:
@swift-ci please ASAN test macOS Platform

gottesmm reviewed (approved):
LGTM. @etcwilde before you commit can you run on swift-ci the as an preset just to make sure it all works in a black box sort of way.
